### DISCUSSION TITLE
Potential Problem with Projectile Launching Position

### DISCUSSION THREAD

**Discussion by 4-S-H** [FakeFishGames/Barotrauma#14611] [Code questions and modding support]
https://github.com/FakeFishGames/Barotrauma/blob/0e8fb6569d2810e2f8ad5fb17b4bba546cc5739a/Barotrauma/BarotraumaShared/SharedSource/Items/Components/Projectile.cs#L383-L413
```
//spawn the projectile body.GetMaxExtent() away from the position where the raycast hit the obstacle
Vector2 newPos = weaponPos - Vector2.Normalize(spawnPos - projectilePos) * Math.Max(Item.body.GetMaxExtent(), 0.1f);
```
On line 398, the comment suggests that the displacement should be subtracted from where the obstacle is, but the code on line 399 seems to be subtracting from `weaponPos`, potentially causing the actual launching position to be much further 'backwards'.

**Comment by itchyOwl** [comment]
Hmm, based on the comment, I'd expect it should be `weaponPos +` the direction towards the barrel by the size of the projectile... However, adjusting `weaponPos` like that causes the projectile to spawn projectile's size times away from the barrel, which is probably not what we want here:
![image](https://github.com/user-attachments/assets/e8a2de44-4390-4fe2-a009-b07791cbd120)

The current code adjusts the position away form the barrel, like this, so that you can't shoot through obstacles by pushing the barrel through them:
![image](https://github.com/user-attachments/assets/fa42af26-5252-41d4-b19c-d4a349966b70)

So it depends which direction _away_ means. I think the code would be more clear if it'd just add to the weapon position the "correct" direction, like so: `Vector2 newPos = weaponPos + Vector2.Normalize(weaponPos - spawnPos)`, but functionally this is the same as `weaponPos - Vector2.Normalize(spawnPos - weaponPos)`.

So in short, I think you are correct that it makes the projectile to be launched further behind, but that seems to be intentional.
